rfcs/0000-ckb-softfork-activation/0000-ckb-softfork-activation.md

```diff
@@ -0,0 +1,132 @@
+
+By restricting the top 3 bits to 001, we get 29 out of those for this proposal and support future upgrades for different mechanisms. When a block `version` does not have top bits 001, it is treated as if all bits are 0 for deployments.
+
+Miners should continue setting the bit in the LOCKED_IN phase, so uptake is visible, though this does not affect consensus rules.
+
+### New consensus rules
+The new consensus rules for each softfork are enforced for each block with an ACTIVE state.
+
+### State transitions
+
+![State transitions](images/state-transitions.png)
+
+The genesis epoch has a state DEFINED for each deployment, by definition.
+
+```rust
+if epoch.is_genesis() {
+    return ThresholdState::DEFINED;
+}
+```
+
+We remain in the initial state until we reach the `start_epoch`.
+
+```rust
+match state {
+    ThresholdState::DEFINED => {
+        if epoch.number() >= start {
+            next_state = ThresholdState::STARTED;
+        }
+    }
+```
+
+After an epoch in the STARTED state, we tally the bits set and transition to LOCKED_IN if a sufficient number of blocks in the past epoch set the deployment bit in their version numbers. If the threshold has not been met and we reach the `timeout_epoch`, we transition directly to FAILED.
+
+Note that a block's state never depends on its version, only on that of its ancestors.
+
+```rust
+match state {
+    ThresholdState::STARTED => {
+        let mut count = 0;
+        for block in (0..epoch_length) {
+            if (block.version() & 0xE0000000 == 0x20000000 && (block.version() >> bit) & 1 == 1) {
+                ++count;
+            }
+        }
+        let threshold_number = threshold_number(epoch_length, threshold);
+        if count >= threshold_number {
+            next_state = ThresholdState::LOCKED_IN;
+        } else if epoch_ext.number() >= timeout {
+            next_state = ThresholdState::FAILED;
+        }
+    }
+```
+
+After at least one epoch of LOCKED_IN, we automatically transition to ACTIVE if the `minimum_activation_epoch` is reached. Otherwise, LOCKED_IN continues.
+
+```rust
+ThresholdState::LOCKED_IN => {
+    if epoch.number() >= min_activation_epoch {
+        next_state = ThresholdState::ACTIVE;
+    }
+}
+```
+
+Furthermore, ACTIVE and FAILED are terminal states in which a deployment stays once reached.
+
+```rust
+ThresholdState::FAILED | ThresholdState::ACTIVE => {
+    // Nothing happens, these are terminal states.
+}
+```
+
+### Warning mechanism
+To support upgrade warnings, an extra "unknown upgrade" is tracked, using the "implicit bit" mask = (version & ~expected_version) != 0. Mask will be non-zero whenever an unexpected bit is set in `version`. Whenever LOCKED_IN for the unknown upgrade is detected, the software should warn loudly about the upcoming softfork. It should warn even more loudly after the next epoch (when the unknown upgrade is in the ACTIVE state).
```

Comment:
Does "the software" mean CKB node or wallet or something else? What does "warn loudly" mean?

> It should warn even more loudly after the next epoch (when the unknown upgrade is in the ACTIVE state).

Why the unknown upgrade is in ACTIVE in the "next epoch" of LOCKED_IN?

Reply:
From my understanding, such an "unknown upgrade" can only triggered by miners setting a bit in `version` when that bit has no corresponding deployment info encoded in the CKB node. If so it can never be ACTIVE because it has no `minimum_activation_epoch' defined.

Reply:
Remove this section for now.